File: python/space-age/space_age.py
class SpaceAge(object):
    base = 31557600
    factors = dict(
        earth=1.0,
        mercury=0.2408467,
        mars=1.8808158,
        jupiter=11.862615,
        saturn=29.447498,
        uranus=84.016846,
        neptune=164.79132,
        venus=0.61519726)

    def __init__(self, seconds):
        self.seconds = seconds

    def __getattr__(self, name):
        planet = name[3:]
        factor = self.factors[planet]
        return lambda: self._years(factor)

    def _years(self, factor):
        return round(self.seconds / (self.base * factor), 2)

File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

ageOn :: Planet -> Float -> Float
ageOn Earth seconds = seconds / 31557600
ageOn planet seconds = ageOn Earth seconds / f
    where f = case planet of Mercury -> 0.2408467
                             Venus -> 0.61519726
                             Mars -> 1.8808158
                             Jupiter -> 11.862615
                             Saturn -> 29.447498
                             Uranus -> 84.016846
                             Neptune -> 164.79132

ageOn' :: Planet -> Float -> Float
ageOn' Earth seconds = seconds / 31557600
ageOn' planet seconds =
    let f = case planet of Mercury -> 0.2408467
                           Venus -> 0.61519726
                           Mars -> 1.8808158
                           Jupiter -> 11.862615
                           Saturn -> 29.447498
                           Uranus -> 84.016846
                           Neptune -> 164.79132
    in ageOn Earth seconds / f
